### PR TITLE
[Backend] As a user, I can logout

### LINKED ISSUE #12: [Backend] As a user, I can logout
### Why?
Revoke the user's session once the he makes a request to logout his account.

### Acceptance Criteria
- [x] Revoke the user's session
- [x] Redirect to the root page

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/12

## What happened 👀

Revoke the user's session once the he makes a request to logout his account.

## Insight 📝

- [x]  Revoke the user's session
- [x]  Redirect to the root page

Extra:

- Create a `dashboard` page for authenticated user can access
- Update policy in `root` page for any user can access
- Re-design `policy` workflow and assign to each action in the controller instead of just only within the controller

Policy Workflow:

controllers/base.go
```go
type BaseController struct {
	actionPolicy map[string]Policy
}

type Policy struct {
	requireAuthenticatedUser bool
	requireGuestUser         bool
}

func (c *BaseController) Prepare() {
        c.actionPolicy = make(map[string]Policy)

        app, ok := c.AppController.(NestPreparer)
	if ok {
		app.NestPrepare()
	}
}

func (c *BaseController) handleAuthorizeRequest() {
       ...
}
```

controllers/session.go
```go
type SessionController struct {
	BaseController
}

func (c *SessionController) NestPrepare() {
	c.actionPolicyMapping()
}

func (c *SessionController) actionPolicyMapping() {
        // All user can access
        c.MappingPolicy("Index", Policy{})

        // Only guest user can access
	c.MappingPolicy("New", Policy{requireGuestUser: true})
	c.MappingPolicy("Create", Policy{requireGuestUser: true})

        // Only authenticated user can access
	c.MappingPolicy("Delete", Policy{requireAuthenticatedUser: true})
}
```

## Proof Of Work 📹

| State of session |
|-----------------|
||
| Before Sign In |
| Screen Shot 2564-02-18 at 14 31 22 |
||
| Signed In |
| Screen Shot 2564-02-18 at 14 31 56 |
||
| After Sign out |
| Screen Shot 2564-02-18 at 14 32 28 |

Demo
Screen Recording 2564-02-18 at 14 24 34

All green with testing
Screen Shot 2564-02-18 at 12 56 04

### CODE REVIEW

File: assets/stylesheets/components/_greeting.scss [hunk 1]
```diff
@@ -0,0 +1,14 @@
+.greeting {
```
Comment (Lahphim) [inline assets/stylesheets/components/_greeting.scss:1]: This will improve in frontend part ✌️

File: controllers/session.go [hunk 2]
```diff
@@ -65,3 +74,30 @@ func (c *SessionController) Create() {
...
+// @Title Delete
+// @Description revoke a session
+// @Success 302 redirect to root path with success message
+// @Failure 302 redirect to current path with error message
+// @router / [get]
+func (c *SessionController) Delete() {
+	flash := web.NewFlash()
+	redirectPath := "/"
+
+	u, err := url.Parse(c.Ctx.Request.Header.Get("Referer"))
```
Comment (carryall) [inline controllers/session.go:88]: @Lahphim Can you point me out what is this for?

File: controllers/base.go [hunk 3]
```diff
@@ -83,3 +107,20 @@ func (c *BaseController) ensureGuestUser() bool {
...
+	c.LayoutSections = make(map[string]string)
+	c.LayoutSections["FlashMessage"] = "shared/alert.html"
+	c.LayoutSections["HeaderContent"] = "shared/header.html"
+}
+
+func (c *BaseController) initActionPolicy() {
+	c.actionPolicy = make(map[string]Policy)
+}
+
+func (c *BaseController) updateCurrentUserData() {
```
Comment (olivierobert) [inline controllers/base.go]: Not sure if it's an update 🤔 Updating would mean to change a value from something to something else. How about using set or assign here? Assign can be better in the sense that it's used in the view 💭 

```suggestion
func (c *BaseController) assignCurrentUser() {
```

**Comment by carryall** [comment]
@Lahphim Should we also update the redirect path after login to `dashboard`?

File: controllers/dashboard.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	c.actionPolicyMapping()
+}
+
+// URLMapping maps dashboard controller actions to functions
+func (c *DashboardController) URLMapping() {
+	c.Mapping("Index", c.Index)
+}
+
+// actionPolicyMapping maps dashboard controller actions to policies
+func (c *DashboardController) actionPolicyMapping() {
```
Comment (olivierobert) [inline controllers/dashboard.go:23]: AT first, I was not sure about the naming. However, now I can see how it's consistent with the previous `URLMapping` method 🌟

File: controllers/dashboard_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+			It("redirects to sign-in page", func() {
+				response := MakeRequest("GET", "/dashboard", nil)
+				currentPath := GetCurrentPath(response)
+
+				Expect(response.StatusCode).To(Equal(http.StatusFound))
+				Expect(currentPath).To(Equal("/user/sign_in"))
+			})
+		})
+
+		Context("when the user has already signed in", func() {
```
Comment (olivierobert) [inline controllers/dashboard_test.go]: In terms of ordering of tests, usually, the happy path is listed before the sad paths 💭

**Review by olivierobert — COMMENTED** [review]
The action policy implementation is a step up from the previous implementation and a good addition to the use of `NestPrepare()` 👍

File: controllers/base.go [hunk 6]
```diff
@@ -60,16 +75,25 @@ func (c *BaseController) GetSessionCurrentUser() (user *models.User) {
...
+	return c.DelSession(CurrentUserKey)
+}
+
 func (c *BaseController) handleAuthorizeRequest() {
-	if c.requireGuestUser && !c.ensureGuestUser() {
+	_, actionName := c.GetControllerAndAction()
+	actionPolicy := c.actionPolicy[actionName]
+
+	if actionPolicy.requireGuestUser && !c.ensureGuestUser() {
 		c.Redirect("/", http.StatusFound)
```
Comment (carryall) [inline controllers/base.go]: @Lahphim Should we redirect to the dashboard? as you said that `dashboard` page is for the authenticated user

File: controllers/session.go (hunk 2 above)
Comment (Lahphim) [reply controllers/session.go:88]: @carryall I just want to improve a little bit with user experience when signing out failed by redirect to the previous path.

For example, I'm in the dashboard page then press sign out and get an error after that somehow.
I should stay at the dashboard page.

`Referer` is auto attached to header to tell us that where we come from.

File: controllers/dashboard_test.go (hunk 5 above)
Comment (Lahphim) [reply controllers/dashboard_test.go]: Updated here https://github.com/Lahphim/go-crawler-challenge/pull/51/commits/ec0ed00ac14f5ca9bcd6cde4d8b644d5f287c5d0

File: controllers/base.go (hunk 3 above)
Comment (Lahphim) [reply controllers/base.go]: updated here https://github.com/Lahphim/go-crawler-challenge/pull/51/commits/ec0ed00ac14f5ca9bcd6cde4d8b644d5f287c5d0

File: controllers/base.go (hunk 6 above)
Comment (Lahphim) [reply controllers/base.go]: updated here https://github.com/Lahphim/go-crawler-challenge/pull/51/commits/e4fea30272cd760e4f45313dd12df21693fc07ea

**Review by carryall — APPROVED** [review]

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**